### PR TITLE
fix(surround_obstacle_checker): prevent crash when all checks for object types to are set to false

### PR DESCRIPTION
**Pull request by danielsanchezaran**
### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at e588cb4</samp>

Refactor and improve `SurroundObstacleCheckerNode` for planning. This change introduces two new member variables to optimize the dynamic object check, fixes a bug with the optional return type, and adds some input validation. The changes affect `node.cpp` and `node.hpp` files.

Refactor `surround_obstacle_checker` node to skip surround check if all checks for object types to false (and throw an Info message accordingly) or if there is no object pointer.

Solves this issue: https://github.com/autowarefoundation/autoware.universe/issues/5642
## Tests performed
in PSim set all checks for object types to false, check if there is a crash or not -> NO crashes with this fix.

https://github.com/autowarefoundation/autoware.universe/assets/25967964/b3207577-68ac-4a6d-b215-be0379d51cdc

## Effects on system behavior

Prevents crashes when all obstacle targets are disabled

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by VRichardJP** [comment]
Isn't static checking supposed to work even if dynamic checking is disabled?

**Comment by danielsanchezaran** [comment]
> Isn't static checking supposed to work even if dynamic checking is disabled?

Thanks for your question. 
I am sorry, but I don't understand the question. The surround checker detects dynamic obstacle types (ie. pedestrian, car, bus ,etc.) but it does not judge if the vehicle should remain stopped based on the obstacle speed, the check is only based on distance between the ego and the closest obstacle. Does this answer your question?

**Comment by VRichardJP** [comment]
I mean the surround obtacle checker also check obstacles based on pointcloud:

https://github.com/autowarefoundation/autoware.universe/blob/d99e56b3968d01233c0367a824dac154154beb9c/planning/surround_obstacle_checker/src/node.cpp#L397C35-L397C65

This part of the code still works even if all targets types are disabled (this is how I currenytly use the node).

With your PR, pointcloud based checking would not run if all targets are disabled.

File: planning/surround_obstacle_checker/src/node.cpp [hunk 1]
```diff
@@ -289,22 +294,17 @@ void SurroundObstacleCheckerNode::onTimer()
   if (node_param_.pointcloud_enable_check && !pointcloud_ptr_) {
     RCLCPP_INFO_THROTTLE(
       this->get_logger(), *this->get_clock(), 5000 /* ms */, "waiting for pointcloud info...");
-    return;
```
Comment (satoshi-ota) [inline planning/surround_obstacle_checker/src/node.cpp:292]: > I mean the surround obtacle checker also check obstacles based on pointcloud:
> 
> https://github.com/autowarefoundation/autoware.universe/blob/d99e56b3968d01233c0367a824dac154154beb9c/planning/surround_obstacle_checker/src/node.cpp#L397C35-L397C65
> 
> This part of the code still works even if all targets types are disabled (this is how I currenytly use the node).
> 
> With your PR, pointcloud based checking would not run anymore if all targets are disabled.

@VRichardJP Could you check latest file changes? I think pointcloud based checking can work even when all dynamic objects flags are `false`.

**Review by satoshi-ota — APPROVED** [review]
It works well in Psim. LGTM :+1:

**Pull request merged**